core/src/main/java/org/apache/iceberg/BaseSnapshot.java

```diff
@@ -52,6 +52,10 @@ class BaseSnapshot implements Snapshot {
   private transient List<DataFile> cachedAdds = null;
   private transient List<DataFile> cachedDeletes = null;
 
+  // lazily initialized in newer API
```

Comment:
I'd prefer not to keep a copy of the FileIO. I think it should remain `final` and we should fail if the wrong method is used after deserialization.

Reply:
Moved it back to final. What should we throw if the wrong method is called? And should we only throw is the cached changes aren't present, or simply throw if the old methods are used _at all_?